core/src/main/kotlin/com/strumenta/kolasu/transformation/DummyNodes.kt

```diff
@@ -0,0 +1,96 @@
+
+/**
+ * This logic instantiate a node of the given class with dummy values.
+ * This is useful because it permits to add an element that "fit" and make the typesystem happy.
+ * Typically, the only goal of the element would be to hold some annotation that indicates that the element
+ * is representing an error or a missing transformation or something of that sort.
+ */
+fun <T : Node> KClass<T>.dummyInstance(): T {
+    val kClassToInstantiate = this.toInstantiableType()
+    val emptyConstructor = kClassToInstantiate.constructors.find { it.parameters.isEmpty() }
+    if (emptyConstructor != null) {
+        return emptyConstructor.call()
+    }
+    val constructor = kClassToInstantiate.primaryConstructor!!
+    val params = mutableMapOf<KParameter, Any?>()
+    constructor.parameters.forEach { param ->
+        val mt = param.type.javaType
+        val value = when {
+            param.type.isMarkedNullable -> null
+            mt is ParameterizedType && mt.rawType == List::class.java -> mutableListOf<Any>()
+            (param.type.classifier as KClass<*>).isSubclassOf(Node::class) ->
+                (param.type.classifier as KClass<out Node>).dummyInstance()
+            param.type == String::class.createType() -> "DUMMY"
+            param.type.classifier == ReferenceByName::class -> ReferenceByName<PossiblyNamed>("UNKNOWN")
+            param.type == Int::class.createType() -> 0
+            param.type == Float::class.createType() -> 0.0f
+            param.type == Long::class.createType() -> 0L
+            param.type == Double::class.createType() -> 0.0
+            param.type == Boolean::class.createType() -> false
+            (param.type.classifier as KClass<*>).isSubclassOf(Enum::class)
+            -> (param.type.classifier as KClass<*>).java.enumConstants[0]
+            else -> TODO("Param type ${param.type}")
+        }
+        params[param] = value
+    }
+    return constructor.callBy(params)
+}
+
+private fun <T : Any> KClass<T>.toInstantiableType(): KClass<out T> {
+    return when {
+        this.isSealed -> {
+            val subclasses = this.sealedSubclasses.filter { it.isDirectlyOrIndirectlyInstantiable() }
+            if (subclasses.isEmpty()) {
+                throw IllegalStateException("$this has no instantiable sealed subclasses")
+            }
+            val subClassWithEmptyParam = subclasses.find { it.constructors.any { it.parameters.isEmpty() } }
+            if (subClassWithEmptyParam == null) {
+                if (subclasses.size == 1) {
+                    subclasses.first().toInstantiableType()
+                } else {
+                    // Some constructs are recursive (think of the ArrayType)
+                    // We either find complex logic to find the ones that aren't or just pick one randomly.
+                    // Eventually we will build a tree
+                    val r = Random.Default
+                    subclasses[r.nextInt(subclasses.size)].toInstantiableType()
```

Comment:
I'm not a fan of random here because parsing the same file twice will likely give different results, which I think is undesirable (even if we can say that the behavior is unspecified for code that we don't yet support).
I would also consider a mechanism (marker interface or annotation) with which we can designate a specific subclass as the default/fallback node, because in my experience it's not uncommon to have classes such as _UnsupportedStatement_ or _ExpressionError_ already in use.

Reply:
That is a good point. My goal is to avoid that we pick always the wrong one as `ArrayType(ArrayType(ArrayType(...)))`, but perhaps what we can do is to pick a pseudo-random number, that actually depends on the depth of the "DummyNode". I will work on that. Adding an annotation could make our life easier here, but it would require extra work when building the language, so I would lean toward doing more work on this code to have to do less work on all the languages

Reply:
I see. Anyway I guess this only applies to 1.5 because in 1.6 we'll have machine-generated error nodes to use.